src/cli/self_update/unix.rs

```diff
@@ -57,66 +57,50 @@ pub fn complete_windows_uninstall() -> Result<utils::ExitCode> {
+            let file = utils::read_file("rcfile", &rc)?;
             let file_bytes = file.into_bytes();
-            let addition_bytes = addition.into_bytes();
-
-            let idx = file_bytes
-                .windows(addition_bytes.len())
-                .position(|w| w == &*addition_bytes);
-            if let Some(i) = idx {
-                let mut new_file_bytes = file_bytes[..i].to_vec();
-                new_file_bytes.extend(&file_bytes[i + addition_bytes.len()..]);
-                let new_file = String::from_utf8(new_file_bytes).unwrap();
-                utils::write_file("rcfile", rcpath, &new_file)?;
-            } else {
-                // Weird case. rcfile no longer needs to be modified?
+            if let Some(idx) = file_bytes
+                .windows(source_bytes.len())
+                .position(|w| w == source_bytes.as_slice())
+            {
+                // Here we rewrite the file without the offending line.
+                let mut new_bytes = file_bytes[..idx].to_vec();
+                new_bytes.extend(&file_bytes[idx + source_bytes.len()..]);
+                let new_file = String::from_utf8(new_bytes).unwrap();
+                utils::write_file("rcfile", &rc, &new_file)?;
             }
-        } else {
-            unreachable!()
         }
     }
 
     Ok(())
 }
 
-pub fn write_env() -> Result<()> {
-    let env_file = utils::cargo_home()?.join("env");
-    let env_str = format!("{}\n", shell_export_string()?);
-    utils::write_file("env", &env_file, &env_str)?;
-    Ok(())
-}
-
-pub fn shell_export_string() -> Result<String> {
-    let path = format!("{}/bin", canonical_cargo_home()?);
-    // The path is *prepended* in case there are system-installed
-    // rustc's that need to be overridden.
-    Ok(format!(r#"export PATH="{}:$PATH""#, path))
-}
+pub fn do_add_to_path() -> Result<()> {
+    let mut scripts = vec![];
 
-pub fn do_add_to_path(methods: &[PathUpdateMethod]) -> Result<()> {
-    for method in methods {
-        if let PathUpdateMethod::RcFile(ref rcpath) = *method {
-            let file = if rcpath.exists() {
-                utils::read_file("rcfile", rcpath)?
-            } else {
-                String::new()
-            };
-            let addition = format!("\n{}", shell_export_string()?);
-            if !file.contains(&addition) {
-                utils::append_file("rcfile", rcpath, &addition).chain_err(|| {
+    for sh in shell::get_available_shells() {
+        let source_cmd = format!("\n{}", sh.source_string()?);
+        for rc in sh.update_rcs() {
+            if !rc.is_file() || !utils::read_file("rcfile", &rc)?.contains(&source_cmd) {
+                utils::append_file("rcfile", &rc, &source_cmd).chain_err(|| {
                     ErrorKind::WritingShellProfile {
-                        path: rcpath.to_path_buf(),
+                        path: rc.to_path_buf(),
                     }
                 })?;
+                let script = sh.env_script();
```

Comment:
I think this code is suffering over-generalism: we know we've created a single env file on disk in the current code, we don't need to track this in a general fashion. Just write the file once, outside the loop (unconditionally, so that the latest version of the file is written).

Reply:
The reason I wrote it this way was in fact planning ahead for and thinking about other shells, yes. It seemed like an inevitable complication.